InstantSearch Insights v2.0

VladislavFitz opened the pull request:
New version of InstantSearch Insights supporting different use cases (Click Analytics/ A/B testing/ Personalisation), new endpoint and events batching.

VladislavFitz committed:
Implemented new models and API call

VladislavFitz committed:
Refactored usecases. Added unit tests

VladislavFitz committed:
Additional unit tests

VladislavFitz committed:
Updated objc tests

VladislavFitz committed:
Added integration tests templates

VladislavFitz committed:
index name is not Insights object dependency anymore. Replaced Filter by raw String

VladislavFitz committed:
Renamed index -> indexName in CoreEvent constructor

VladislavFitz committed:
Decoupled Insights instance from index name. Updated documentation

VladislavFitz committed:
Transform time interval to int while encoding event

VladislavFitz committed:
EventSynchronizer: implemented queuing, added switch for local storage

VladislavFitz committed:
Added convenient delete function to local storage

VladislavFitz committed:
Replaced index by appId

VladislavFitz committed:
Implemented integration tests

VladislavFitz committed:
Added documentation

VladislavFitz committed:
Replaced Operation queue by injectable DispatchQueue in EventSynchronizer

VladislavFitz committed:
Renamed EventsSynchronizer -> EventsController
Moved flushTimer to main RunLoop

VladislavFitz committed:
Upgraded to swift 4.2

VladislavFitz committed:
Added warning in documentation for shared computed variable

VladislavFitz committed:
Insights test structure improvement

VladislavFitz committed:
Added API endpoint region support

VladislavFitz committed:
Renamed EventsSynchronizer -> EventsController in test

VladislavFitz committed:
Removed unnecessary queryID reference field in Personalisation

VladislavFitz committed:
Made region variable public and settable from objc

VladislavFitz committed:
Replaced exception throwing by logging

VladislavFitz committed:
Moved logger output to main thread

VladislavFitz committed:
Increased expectations wait time for being sure that the test suite pass on bitrise

spinach commented on Insights/ABTesting.swift:
```diff
@@ -0,0 +1,105 @@
+import Foundation
+
+/**
+ A/B Testing allows you to create 2 alternative indices, A and B, each with their own settings, and to put them both live, to see which one performs best. Capture the same user events for both A and B. Measure these captured events against each other, creating scores. Use these scores to determine whether A or B is a better user experience. Adjust your main index accordingly.
```
Can we line wrap this? line is too long and we should avoid horizontal scrolling.

Note that we should add Swiftlint actually here to solve these kind of issues. 

To add the config folder that has swiftlint config file in it, you can follow [this step](https://github.com/algolia/instantsearch-ios-config#to-add-the-config-folder-in-a-brand-new-project-in) in the README

Then, in Xcode build phases, add a run script with the following:
```
if which swiftlint >/dev/null; then
config/swiftlint --config config/.swiftlint.yml
else
echo "warning: SwiftLint not installed, download from https://github.com/realm/SwiftLint"
fi
```

spinach commented on Insights/Event/Click.swift:
```diff
@@ -0,0 +1,98 @@
+
+import Foundation
+
+public struct Click: Event {
```
For public classes/structs, we should put some comments to explain.

If we want to be "perfect", we should document all public properties/methods. but no need to go into too much details as well

spinach commented on Insights/Extensions/Decodable.swift:
```diff
@@ -0,0 +1,9 @@
+//  Copyright © 2018 Algolia. All rights reserved.
+//
+
+import Foundation
```
Is this empty file still needed?

spinach commented on Insights/EventProcessor.swift:
```diff
@@ -0,0 +1,15 @@
+
+import Foundation
+
+protocol EventProcessor: class {
```
I would go for EventProcessable 

"protocols that describe a capability should end in -able or -ible"

spinach commented on Insights/ABTesting.swift:
```diff
@@ -0,0 +1,105 @@
+        self.logger = logger
+    }
+    
+    // Only for Objective-C
```
There's a trick that we use in the client for objective-c only methods: 

The Swift name is prefixed with z_objc_, which makes it obvious that it is intended for Objective-C, and also ensures it appears last in code completion.
Using an @objc(name) annotation, we map this property to the parameter’s name. Since the original property is not visible in Objective-C, no conflict is induced by this. All goes as if the property simply had a different type in Objective-C.
The fallback property is not documented, which means it does not appear in the generated reference documentation.

More details here https://blog.algolia.com/supporting-swift-objective-c/ if you want to read about it

VladislavFitz committed:
All the event models are internal now. Introduced CoreEventContainer protocol to reduce boilerplate.

VladislavFitz committed:
Removed obsolete files

VladislavFitz committed:
Removed obsolete file, added source/test files to all the targets

VladislavFitz committed:
Integration tests fix

VladislavFitz committed:
Test fix

spinach commented on Insights/EventsController.swift:
```diff
@@ -0,0 +1,124 @@
+        self.flushTimer = flushTimer
+
+        RunLoop.main.add(flushTimer, forMode: .default)
+        deserialize()
```
we should put more info into that. reading deserialize hides a lot of information. what is it deserializing?

- I would clarify in the name of the function what it is deserializing
- potentially, if it's mutating a property of the class (in this case `self.eventsPackage`), then for readability, I would prefer when calling the method to do `self.eventsPackage = deserialize()` or `try deserialize` or something something like that.

spinach commented on Insights/EventsController.swift:
```diff
@@ -0,0 +1,124 @@
+        flush(eventsPackages)
+    }
+    
+    func flush(_ eventsPackages: [EventsPackage]) {
```
for consistency with the method sync below, can we have named parameter for this method as well?

spinach commented on Insights/EventProcessor.swift:
```diff
@@ -0,0 +1,15 @@
+//
```
Any reason why this protocol isn't in the protocol folder? also this comment applies for a few other files that are not under a folder/group

spinach commented on Insights/Personalization.swift:
```diff
@@ -0,0 +1,189 @@
+//
```
I would put this file + ClickAnalytics + ABTest under one folder/group

spinach commented on Insights/Insights.swift:
```diff
@@ -10,174 +10,88 @@ import Foundation
-      logger.debug(message: "Error reading a file for \(credentials.indexName)")
-      return
+    
+    public var loggingEnabled: Bool = false {
```
A small comment to explain what gets logged exactly

spinach commented on Insights/Insights.swift:
```diff
@@ -10,174 +10,88 @@ import Foundation
+    private let eventsController: EventsController
+    public let personalization: Personalization
+    public let abTesting: ABTesting
+    public let clickAnalytics: ClickAnalytics
```
Some comments for the 3 above

spinach commented on Insights/Insights.swift:
```diff
@@ -10,174 +10,88 @@ import Foundation
-public enum InsightsException: Error {
-  case credentialsNotFound(String)
+    
+    private let eventsController: EventsController
```
The type can be the abstract `EventsProcessor`. That can be helpful for DI in case we want to mock that later on

spinach commented on Insights/EventsController.swift:
```diff
@@ -0,0 +1,124 @@
+//
+//  EventsController.swift
```
If we rename the protocol from `EventProcessor` to `EventProcessable`, then I would name this class `EventProcessor `

PLNech commented on InsightsTests/Unit/InsightsTests.swift:
```diff
@@ -0,0 +1,101 @@
+        XCTAssertNil(insightsRegister)
+        
+    }
+    
```
What about a `testInitWithoutSpecifyingOneOfTheTwoApps` that would confirm exception is thrown when several apps & missing a parameter?

VladislavFitz committed:
Wrapped usecase documentation

VladislavFitz committed:
Squashed 'config/' content from commit 75f49d3

git-subtree-dir: config
git-subtree-split: 75f49d32bd59960c3458658b7e6d0f48df7fcaf0

VladislavFitz committed:
Merge commit '0e34accb3a5faa914df2381d46520257fef07b4a' as 'config'

VladislavFitz committed:
Added swift lint. Fixed lint warnings.

VladislavFitz committed:
Improved objc bridging

VladislavFitz committed:
Replaced CA & AB testing & Personalisation by  Search & Visit. Added one objectID overloads for all the analytics functions accepting object IDs.

VladislavFitz committed:
Added test coverage for single object ID overloads

VladislavFitz committed:
EventsPackage collection conformance

VladislavFitz committed:
Simplified and improved naming in EventsProcessor

VladislavFitz committed:
Generalized WebService sync() method signature. Updated doc.

VladislavFitz committed:
Code style and doc improvements

VladislavFitz committed:
Added test for insights "shared" accessor in case of multiple applications registered

VladislavFitz committed:
Added more comments

VladislavFitz committed:
Added Insights opt-in/opt-out functionality + tests

VladislavFitz committed:
Out-of-box user token propagation + tests

VladislavFitz committed:
Completed all targets with missing source files

VladislavFitz committed:
Set safe expectation timeout

VladislavFitz committed:
Fixed test compilation error

VladislavFitz committed:
Missing source files for targets

VladislavFitz committed:
Unit tests for encoding/decoding of ObjectIDsOrFilters

VladislavFitz committed:
Implemented per-app region settings.

VladislavFitz committed:
Decode/encode queryID and positions only if they are present in json/object

spinach commented on InsightsTests/Unit/APIEndpointTests.swift:
```diff
@@ -0,0 +1,42 @@
+    func testURLWithRegion() {
+        
+        let deURL = API.baseURL(forRegion: .de)
+        XCTAssertEqual(deURL, URL(string: "https://insights.de.algolia.io"))
```
Nitpicky: can you extract the string as a variable, since it's being used 5 times? just in case the url changes for some reason in the future

spinach commented on InsightsTests/Integration/WebServiceIntegrationTests.swift:
```diff
@@ -0,0 +1,168 @@
+
+class WebServiceIntegrationTests: XCTestCase {
+
+    let appId = "932LAAGOT3"
```
Extract into env variables as we discussed

spinach commented on InsightsTests/Unit/EventsProcessorTests.swift:
```diff
@@ -0,0 +1,148 @@
+        }
+        
+        let queue = DispatchQueue(label: "test queue", qos: .default)
+        let credentials = Credentials(appId: appId, apiKey: "APIKEY")
```
Nit: move "APIKEY" to a variable like you did for appId

VladislavFitz committed:
Replaced seconds by milliseconds for all events timestamps

VladislavFitz committed:
Removed sensitive data from tests, fixed tests issues

VladislavFitz committed:
Added method chaining prototype playground

VladislavFitz committed:
Added web service completion count tests

VladislavFitz committed:
Reading env variables from info plist

VladislavFitz committed:
Added Date+milliseconds to all the targets

VladislavFitz committed:
Fixed tvOS tests

VladislavFitz committed:
Updated fastlane

VladislavFitz committed:
Fixed env variable getter to make local tests pass

VladislavFitz committed:
Implemented event tracking directly from Insights instance

robertmogos commented on Insights/Usecases/Visit.swift:
```diff
@@ -0,0 +1,150 @@
+    func view(eventName: String,
+              indexName: String,
+              userToken: String? = .none,
+              timestamp: Int64 = Date().millisecondsSince1970,
```
Based on the last week problems and conversation, we should not default the timestamp. The server accepts optional timestamp

robertmogos replied:
Same for all the other methods

VladislavFitz committed:
Polished APIs and added mock tests for Insights

VladislavFitz committed:
Fixed parameters order in documentation

VladislavFitz committed:
Replaced Int64 in ms with TimeInterval in s for public functions. Updated readme

VladislavFitz committed:
Fixed documentation by replacing ms with seconds.
Moved optional parameters to the end of methods for better conformance with API on another platforms.

VladislavFitz committed:
Removed objc method from SearchEventTrackable protocol.
Added objc tracking method overloads with reduced optional parameters.

VladislavFitz committed:
Added convenient objc overloads for all the platforms

VladislavFitz committed:
Updated readme layout

VladislavFitz committed:
Fixed indentation in readme

VladislavFitz committed:
Set Swift 4.2 in podspec

VladislavFitz committed:
Updated version

VladislavFitz committed:
Removed unexpected event while library initialization. Improved logging

VladislavFitz committed:
EventWrapper logging improvement

VladislavFitz committed:
Updated API & tests after the last discussion.

VladislavFitz committed:
Merge branch 'master' into feat/perso-reco

VladislavFitz committed:
Code style and naming improvements

VladislavFitz committed:
Updated podspec

VladislavFitz committed:
Separated timer controller. Made flushDelay settable.

VladislavFitz committed:
Fixed tests

VladislavFitz committed:
Updated readme

VladislavFitz committed:
Updated podspec

VladislavFitz committed:
Changed track method names to past form

VladislavFitz committed:
Updated readme

VladislavFitz committed:
Updated podspec

VladislavFitz committed:
Updated readme

VladislavFitz committed:
Updated readme with new methods names

VladislavFitz committed:
Changed default flush delay to 30 minutes

VladislavFitz committed:
Added info about customising flush delay and region to readme
